src/ring_theory/tensor_product.lean

```diff
@@ -845,3 +845,75 @@ begin
   rw [←E1.range_val, ←E2.range_val, ←algebra.tensor_product.product_map_range],
   exact (algebra.tensor_product.product_map E1.val E2.val).to_linear_map.finite_dimensional_range,
 end
+
+namespace tensor_product.algebra
+
+variables {R A B M : Type*}
+variables [comm_semiring R] [add_comm_monoid M] [module R M]
+variables [semiring A] [semiring B] [module A M] [module B M]
+variables [algebra R A] [algebra R B]
+variables [is_scalar_tower R A M] [is_scalar_tower R B M]
+
+/-- An auxiliary definition, used for constructing the `module (A ⊗[R] B) M` in
+`tensor_product.algebra.module` below. -/
+def module_aux : A ⊗[R] B →ₗ[R] M →ₗ[R] M :=
+tensor_product.lift
+{ to_fun := λ a, a • (algebra.lsmul R M : B →ₐ[R] module.End R M).to_linear_map,
+  map_add' := λ r t, by { ext, simp only [add_smul, linear_map.add_apply] },
+  map_smul' := λ n r, by { ext, simp only [ring_hom.id_apply, linear_map.smul_apply, smul_assoc] } }
+
+lemma module_aux_apply (a : A) (b : B) (m : M) :
+  module_aux (a ⊗ₜ[R] b) m = a • b • m :=
+by simp [module_aux]
+
+variables [smul_comm_class A B M]
+
+/-- If `M` is a representation of two different `R`-algebras `A` and `B` whose actions commute,
+then it is a representation the `R`-algebra `A ⊗[R] B`.
+
+An important example arises from a semiring `S`; allowing `S` to act on itself via left and right
+multiplication, the roles of `R`, `A`, `B`, `M` are played by `ℕ`, `S`, `Sᵐᵒᵖ`, `S`. This example
+is important because a submodule of `S` as a `module` over `S ⊗[ℕ] Sᵐᵒᵖ` is a two-sided ideal.
+
+NB: This is not an instance because in the case `B = A` and `M = A ⊗[R] A` we would have a diamond
+of `smul` actions. Furthermore, this would not be a mere definitional diamond but a true
+mathematical diamond in which `A ⊗[R] A` had two distinct scalar actions on itself: one from its
+multiplication, and one from this would-be instance. Arguably we could live with this but in any
```

Comment:
Could you add an example to demonstrate this in `tests/instance_diamonds.lean`?

Reply:
You mean a test to demonstrate the would-be diamond!?

Reply:
Yes, one that demonstrates that `this.module : module (A ⊗[R] A) (A ⊗[R] A) \ne semiring.to_module`, for any choice of `R` or `A` that makes the proof easy.

Reply:
I'm thinking something in the style of

https://github.com/leanprover-community/mathlib/blob/8086825f39e99eadd5de30f5ad8c64b3a11dc32b/test/instance_diamonds.lean#L159-L166

Reply:
OK I'll give it a try (though possibly tomorrow as I'm running low on time today).

Not to push back on the valuable code review that you so generously provide in such abundance but what does a test like this give us?

Reply:
Mainly it gives a proof that the things we claim in comments are actually true; it also gives someone doing a redesign in future an example of the type of things they need to think about.

Reply:
I'm out of time for today so I'll have to pick this up tomorrow. I think it will be quite annoying to add this test. How would you feel about adding the following instead:
```lean
variables (R A : Type) [comm_semiring R] [comm_semiring A] [algebra R A]
example (a b c d : A) :
  ((@tensor_product.algebra.module R A A (A ⊗[R] A) _ _ _ _ _ _ _ _ _ _ _ _).smul : A ⊗[R] A → A ⊗[R] A → A ⊗[R] A)
  (a ⊗ₜ b) (c ⊗ₜ d) = (a*b*c) ⊗ₜ d :=
begin
  rw [tensor_product.algebra.smul_def, mul_assoc],
  refl,
end
```
which I think is a clearer way to illustrate the issue.

Reply:
I might have time to address this tomorrow afternoon, if you don't

Reply:
I added a test but though for reasons I do not understand, it times out when I place it in `test/instance_diamonds.lean` so I have commented it out.

This is disappointing but I'm now going to appeal to Johan's delegation and merge.